server/controller/src/main/java/ai/starwhale/mlops/domain/model/ModelService.java

```diff
@@ -428,18 +439,49 @@ public void pull(String projectUrl, String modelUrl, String versionUrl, HttpServ
             log.error("listing file from storage failed {}", modelVersionEntity.getStoragePath(), e);
             throw new SwProcessException(ErrorType.STORAGE);
         }
+
         if (CollectionUtils.isEmpty(files)) {
             throw new SwValidationException(ValidSubject.MODEL, "model version empty folder");
         }
-        String filePath = files.get(0);
-        try (InputStream fileInputStream = storageAccessService.get(filePath);
-                ServletOutputStream outputStream = httpResponse.getOutputStream()) {
-            long length = fileInputStream.transferTo(outputStream);
-            String fileName = filePath.substring(modelVersionEntity.getStoragePath().length() + 1);
-            httpResponse.addHeader("Content-Disposition", "attachment; filename=\"" + fileName + "\"");
+
+        try (ServletOutputStream outputStream = httpResponse.getOutputStream()) {
+            long length = 0;
+            if (files.size() == 1 && files.get(0).endsWith(".swmp")) {
+                var filePath = files.get(0);
+                try (LengthAbleInputStream fileInputStream = storageAccessService.get(filePath)) {
+                    length += fileInputStream.transferTo(outputStream);
+                }
+            } else {
+                try (ArchiveOutputStream archiveOutputStream =
+                         new ArchiveStreamFactory().createArchiveOutputStream(ArchiveStreamFactory.TAR, outputStream)) {
+                    // tar all of these files
+                    for (String filePath : files) {
+                        try (LengthAbleInputStream fileInputStream = storageAccessService.get(filePath)) {
+                            TarArchiveEntry entry = new TarArchiveEntry(new File(
+                                    filePath.substring(modelVersionEntity.getStoragePath().length() + 1)));
+                            if (entry.isDirectory()) {
```

Comment:
directories should not be ignored

Reply:
Is it necessary to keep the empty directory